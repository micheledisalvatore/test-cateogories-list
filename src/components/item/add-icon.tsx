import React from 'react'

export default () => (
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
    <defs/>
    <path d="M256 0C114.84 0 0 114.84 0 256s114.84 256 256 256 256-114.84 256-256S397.16 0 256 0zm0 475.429c-120.997 0-219.429-98.432-219.429-219.429S135.003 36.571 256 36.571 475.429 135.003 475.429 256 376.997 475.429 256 475.429z"/>
    <path d="M256 134.095c-10.1 0-18.286 8.186-18.286 18.286v207.238c0 10.1 8.186 18.286 18.286 18.286 10.1 0 18.286-8.186 18.286-18.286V152.381c0-10.1-8.186-18.286-18.286-18.286z"/>
    <path d="M359.619 237.714H152.381c-10.1 0-18.286 8.186-18.286 18.286 0 10.1 8.186 18.286 18.286 18.286h207.238c10.1 0 18.286-8.186 18.286-18.286 0-10.1-8.186-18.286-18.286-18.286z"/>
  </svg>
);