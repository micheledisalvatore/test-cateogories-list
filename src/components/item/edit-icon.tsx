import React from 'react';

export default () => (
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="-1 0 393 393.15442">
    <defs/>
    <path d="M342.402344 14.621094c-19.542969-19.496094-51.171875-19.496094-70.710938 0L34.601562 251.710938c-2.886718 2.886718-3.726562 7.242187-2.121093 11L.429688 380.53125c-.945313 3.472656.050781 7.1875 2.605468 9.722656 2.558594 2.535156 6.28125 3.5 9.746094 2.527344l116.933594-32.867188c3.746094 1.574219 8.070312.722657 10.941406-2.148437l237.089844-237.089844c19.523437-19.527343 19.523437-51.183593 0-70.710937zm-84.601563 42.167968l77.351563 78.195313-201.570313 201.566406L55.8125 258.78125zM48.554688 279.808594l64.121093 64.117187-88.316406 24.828125zm315.046874-173.273438l-14.304687 14.304688-77.351563-78.191406 13.886719-13.882813c11.726563-11.699219 30.703125-11.699219 42.429688 0l35.34375 35.34375c11.695312 11.722656 11.695312 30.703125-.003907 42.425781zm0 0"/>
  </svg>
);
