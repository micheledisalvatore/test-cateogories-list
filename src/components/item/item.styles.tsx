import styled, { css } from 'styled-components';

type ContainerProps = {
  readonly isDirectory: boolean;
};

export const Container = styled.li`
  list-style: none;
`;

export const Form = styled.form<ContainerProps>`
  display: flex;
  align-items: center;
  justify-content: space-between;

  ${({ isDirectory }) => isDirectory && css`
    &::before {
      content: ' ';
      display: inline-block;

      border-top: 5px solid transparent;
      border-bottom: 5px solid transparent;
      border-left: 5px solid currentColor;

      vertical-align: middle;
      margin-right: .7rem;
      transition: transform .2s ease-out;

      transform: rotate(90deg);
    }
  `}
`

export const Label = styled.div`
  font-size: 1rem;
  flex-grow: 1;
  text-align: left;
`;

export const Input = styled(Label)`
  border: none;
  padding: 0;
`;

export const Icons = styled.div``;
export const Icon = styled.button`
  background: none;
  border: none;
  margin: 0.2rem;
  cursor: pointer;

  & svg {
    width: 1rem;
    height: 1rem;
  }

  &:hover svg {
    fill: red;
  }
`;
