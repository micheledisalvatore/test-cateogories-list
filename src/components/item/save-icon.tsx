import React from 'react'

export default () => (
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="-6 0 443 443.00624">
    <defs/>
    <path d="M9.863281 443.007812h410.28125c5.519531 0 9.859375-4.980468 9.859375-10.5V133.734375c.023438-2.332031-.773437-4.597656-2.25-6.402344l-88.023437-105.6875c-1.882813-2.289062-4.683594-3.621093-7.648438-3.636719h-32.078125v-8.5c.09375-2.546874-.875-5.019531-2.679687-6.824218-1.800781-1.800782-4.273438-2.7695315-6.820313-2.6757815h-221c-5.523437 0-10.5 3.9765625-10.5 9.4999995v8.5H9.863281C7.21875 18.03125 4.691406 19.109375 2.839844 21 .992188 22.890625-.03125 25.441406.00390625 28.089844v404.417968c0 5.519532 4.33984375 10.5 9.85937475 10.5zm69.140625-423h201v79h-201zm-59 18h39v71.441407c0 5.523437 4.976563 9.558593 10.5 9.558593h221c2.554688.078126 5.027344-.902343 6.828125-2.714843s2.769531-4.289063 2.671875-6.84375V38.007812h27.398438l82.601562 99.347657v285.652343h-390zm0 0"/>
    <path d="M350.003906 209.007812c0-5.523437-4.476562-10-10-10h-251c-5.523437 0-10 4.476563-10 10v164c0 5.519532 4.476563 10 10 10h251c5.523438 0 10-4.480468 10-10zm-20 154h-231v-144h231zm0 0M240.003906 39.507812c-5.523437 0-10 4.476563-10 10v19.941407c0 5.523437 4.476563 10 10 10 5.523438 0 10-4.476563 10-10V49.507812c0-5.523437-4.476562-10-10-10zm0 0M210.003906 39.507812c-5.523437 0-10 4.476563-10 10v19.941407c0 5.523437 4.476563 10 10 10 5.523438 0 10-4.476563 10-10V49.507812c0-5.523437-4.476562-10-10-10zm0 0"/>
    <path d="M129.503906 264.007812h81c5.523438 0 10-4.480468 10-10 0-5.523437-4.476562-10-10-10h-81c-5.523437 0-10 4.476563-10 10 0 5.519532 4.476563 10 10 10zm0 0M129.503906 304.007812h171c5.523438 0 10-4.480468 10-10 0-5.523437-4.476562-10-10-10h-171c-5.523437 0-10 4.476563-10 10 0 5.519532 4.476563 10 10 10zm0 0M129.503906 344.007812h171c5.523438 0 10-4.480468 10-10 0-5.523437-4.476562-10-10-10h-171c-5.523437 0-10 4.476563-10 10 0 5.519532 4.476563 10 10 10zm0 0"/>
  </svg>
);