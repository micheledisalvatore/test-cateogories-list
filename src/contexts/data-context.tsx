import React, { useState, createContext } from 'react'

type Datum = {
  id: number;
  parent: number;
  isDirectory: boolean;
  label: string;
};

const defaultData: Datum[] = [
  {
    label: 'Everything',
    isDirectory: true,
    id: 0,
    parent: -1,
  },
]

type DataContextProps = {
  data: Datum[];
  saveItem: Function;
  deleteItem: Function;
  addItem: Function;
}

export const dataContext = createContext<DataContextProps>({
  data: [],
  saveItem: () => {},
  deleteItem: () => {},
  addItem: () => {},
})

type DataProviderProps = {
  children: JSX.Element;
}

export const DataProvider = ({ children }: DataProviderProps) => {
  const [data, setData] = useState(defaultData)

  const getItemIndexById = (id: number) => data.findIndex(((datum) => datum.id === id));

  const saveItem = (id: number, label: string) => {
    const index = getItemIndexById(id)
    data[index].label = label
    setData([...data])
  }
  const deleteItem = (id: number) => {
    const index = getItemIndexById(id)
    data.splice(index, 1)

    data.forEach((datum, childIndex) => {
      if (datum.parent === id) {
        data.splice(childIndex, 1)
      }
    })

    setData([...data])
  }
  const addItem = (parentId: number, isDirectory: boolean = false) => {
    data.push({
      label: '',
      isDirectory,
      id: Math.floor(Math.random() * 10000),
      parent: parentId,
    })

    setData([...data])
  }

  const value: DataContextProps = {
    data,
    saveItem,
    deleteItem,
    addItem,
  }

  return (
    <dataContext.Provider value={value}>
      { children }
    </dataContext.Provider>
  )
}