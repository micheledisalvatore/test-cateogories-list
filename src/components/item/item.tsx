import React, { useRef, useState, useEffect, useContext } from 'react';

import { dataContext } from '../../contexts/data-context'

import { Container, Label, Input, Icons, Icon, Form } from './item.styles';
import EditIcon from './edit-icon'
import AddIcon from './add-icon'
import DirectoryIcon from './directory-icon'
import DeleteIcon from './delete-icon'
import SaveIcon from './save-icon'

type ItemProps = {
  id: number;
  isDirectory: boolean;
  children: string;
}

export const Item = ({ id, children, isDirectory }: ItemProps) => {
  const [isEditing, setEditingMode] = useState(children === '')
  const input = useRef<HTMLInputElement>(null)
  const { saveItem, deleteItem, addItem } = useContext(dataContext)

  const toggleEdit = () => {
    setEditingMode(!isEditing)
  }

  useEffect(() => {
    if (isEditing && input && input.current) {
      input.current.focus()
    }
  }, [isEditing])

  const handleSave = () => {
    let value: string = ''
    if (isEditing && input && input.current) {
      ({ value } = input.current)
    }
    saveItem(id, value)
    toggleEdit()
  }

  const handleDelete = () => {
    deleteItem(id)
  }

  const handleAdd = (isAddingDirectory: boolean) => () => {
    addItem(id, isAddingDirectory)
  }

  return (
    <Container>
      <Form isDirectory={isDirectory} onSubmit={handleSave} data-testid="form">
        {!isEditing && (
          <Label data-testid="label">
            { children }
          </Label>
        )}
        {isEditing && <Input as="input" type="text" defaultValue={children} ref={input} data-testid="input" />}
        <Icons>
          {!isEditing && (<Icon type="button" onClick={toggleEdit} data-testid="edit"><EditIcon /></Icon>)}
          {isEditing && (<Icon type="submit" onClick={handleSave} data-testid="save"><SaveIcon /></Icon>)}
          {isDirectory && (<Icon type="button" onClick={handleAdd(false)} data-testid="add-item"><AddIcon /></Icon>)}
          {isDirectory && (<Icon type="button" onClick={handleAdd(true)} data-testid="add-category"><DirectoryIcon /></Icon>)}
          <Icon type="button" onClick={handleDelete} data-testid="delete"><DeleteIcon /></Icon>
        </Icons>
      </Form>
    </Container>
  )
}
