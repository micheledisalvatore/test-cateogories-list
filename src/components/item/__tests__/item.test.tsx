import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Item } from '../item';
import { dataContext} from '../../../contexts/data-context'

test('renders the basic component', () => {
  const { container } = render(<dataContext.Provider value={{
    saveItem: jest.fn(),
    deleteItem: jest.fn(),
    addItem: jest.fn(),
  }}><Item id={0} children="foo" isDirectory={false} /></dataContext.Provider>);
  expect(container).toMatchSnapshot()
});

test('emulates the editing feature', () => {
  const saveItem = jest.fn()
  const deleteItem = jest.fn()
  const addItem = jest.fn()

  render(<dataContext.Provider value={{
    saveItem,
    deleteItem,
    addItem,
  }}><Item id={0} children="foo" isDirectory={false} /></dataContext.Provider>);
  expect(screen.queryByTestId('label')).toBeInTheDocument()

  fireEvent.click(screen.queryByTestId('edit'))
  expect(screen.queryByTestId('label')).not.toBeInTheDocument()

  fireEvent.change(screen.queryByTestId('input'), { value: 'foo' })
  fireEvent.submit(screen.queryByTestId('form'))
  expect(saveItem).toBeCalledWith(0, 'foo')
});

test('emulates adding an item and category', () => {
  const saveItem = jest.fn()
  const deleteItem = jest.fn()
  const addItem = jest.fn()

  render(<dataContext.Provider value={{
    saveItem,
    deleteItem,
    addItem,
  }}><Item id={0} children="foo" isDirectory={true} /></dataContext.Provider>);
  expect(screen.queryByTestId('label')).toBeInTheDocument()

  fireEvent.click(screen.queryByTestId('add-item'))
  expect(addItem).toBeCalledWith(0, false)

  fireEvent.click(screen.queryByTestId('add-category'))
  expect(addItem).toBeCalledWith(0, true)
});

test('emulates deleting an item', () => {
  const saveItem = jest.fn()
  const deleteItem = jest.fn()
  const addItem = jest.fn()

  render(<dataContext.Provider value={{
    saveItem,
    deleteItem,
    addItem,
  }}><Item id={0} children="foo" isDirectory={true} /></dataContext.Provider>);
  expect(screen.queryByTestId('label')).toBeInTheDocument()

  fireEvent.click(screen.queryByTestId('delete'))
  expect(deleteItem).toBeCalledWith(0)
});
