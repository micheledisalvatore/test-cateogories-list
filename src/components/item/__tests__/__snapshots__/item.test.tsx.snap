// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the basic component 1`] = `
<div>
  <li
    class="sc-AxjAm fsIKLa"
  >
    <form
      class="sc-AxirZ jbxkeR"
      data-testid="form"
    >
      <div
        class="sc-AxiKw cXlGAn"
        data-testid="label"
      >
        foo
      </div>
      <div
        class="sc-AxhUy kMyssT"
      >
        <button
          class="sc-AxgMl hFnGSe"
          data-testid="edit"
          type="button"
        >
          <svg
            viewBox="-1 0 393 393.15442"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs />
            <path
              d="M342.402344 14.621094c-19.542969-19.496094-51.171875-19.496094-70.710938 0L34.601562 251.710938c-2.886718 2.886718-3.726562 7.242187-2.121093 11L.429688 380.53125c-.945313 3.472656.050781 7.1875 2.605468 9.722656 2.558594 2.535156 6.28125 3.5 9.746094 2.527344l116.933594-32.867188c3.746094 1.574219 8.070312.722657 10.941406-2.148437l237.089844-237.089844c19.523437-19.527343 19.523437-51.183593 0-70.710937zm-84.601563 42.167968l77.351563 78.195313-201.570313 201.566406L55.8125 258.78125zM48.554688 279.808594l64.121093 64.117187-88.316406 24.828125zm315.046874-173.273438l-14.304687 14.304688-77.351563-78.191406 13.886719-13.882813c11.726563-11.699219 30.703125-11.699219 42.429688 0l35.34375 35.34375c11.695312 11.722656 11.695312 30.703125-.003907 42.425781zm0 0"
            />
          </svg>
        </button>
        <button
          class="sc-AxgMl hFnGSe"
          data-testid="delete"
          type="button"
        >
          <svg
            viewBox="-40 0 427 427.00131"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs />
            <path
              d="M232.398438 154.703125c-5.523438 0-10 4.476563-10 10v189c0 5.519531 4.476562 10 10 10 5.523437 0 10-4.480469 10-10v-189c0-5.523437-4.476563-10-10-10zm0 0M114.398438 154.703125c-5.523438 0-10 4.476563-10 10v189c0 5.519531 4.476562 10 10 10 5.523437 0 10-4.480469 10-10v-189c0-5.523437-4.476563-10-10-10zm0 0"
            />
            <path
              d="M28.398438 127.121094V373.5c0 14.5625 5.339843 28.238281 14.667968 38.050781C52.351562 421.390625 65.273438 426.976562 78.796875 427H268c13.527344-.023438 26.449219-5.609375 35.730469-15.449219 9.328125-9.8125 14.667969-23.488281 14.667969-38.050781V127.121094c18.542968-4.921875 30.558593-22.835938 28.078124-41.863282C343.992188 66.234375 327.785156 52.003906 308.597656 52h-51.199218V39.5c.058593-10.511719-4.097657-20.605469-11.539063-28.03125C238.417969 4.046875 228.308594-.0859375 217.796875 0H129c-10.511719-.0859375-20.621094 4.046875-28.0625 11.46875C93.496094 18.894531 89.339844 28.988281 89.398438 39.5V52H38.199219c-19.1875.003906-35.394531 14.234375-37.878907 33.257812-2.480468 19.027344 9.535157 36.941407 28.078126 41.863282zM268 407H78.796875c-17.097656 0-30.398437-14.6875-30.398437-33.5V128h250v245.5c0 18.8125-13.300782 33.5-30.398438 33.5zM109.398438 39.5c-.066407-5.207031 1.980468-10.21875 5.675781-13.894531C118.765625 21.929688 123.789062 19.910156 129 20h88.796875c5.210937-.089844 10.234375 1.929688 13.925781 5.605469 3.695313 3.671875 5.742188 8.6875 5.675782 13.894531V52h-128zM38.199219 72h270.398437c9.941406 0 18 8.058594 18 18s-8.058594 18-18 18H38.199219c-9.941407 0-18-8.058594-18-18s8.058593-18 18-18zm0 0"
            />
            <path
              d="M173.398438 154.703125c-5.523438 0-10 4.476563-10 10v189c0 5.519531 4.476562 10 10 10 5.523437 0 10-4.480469 10-10v-189c0-5.523437-4.476563-10-10-10zm0 0"
            />
          </svg>
        </button>
      </div>
    </form>
  </li>
</div>
`;
